### PR TITLE
Select widget

### PR DESCRIPTION
**Pull request by willmcgugan**
- Adds a Select widget.
- Adds two new CSS rules. `overlay:` and `constrain:`.
- Allows bindings to be set with two values.
- Adds `Region.translate_inside`

Setting `overlay: screen` display a widget above everything else. Like there is an implicit top level layer.

Setting `constrain:` to `x`, `y`, or `both` ensures that the widget doesn't get clipped of the edge of the screen.

This two rules are undocumented as they are experimental. They simplify the implementation of pop-up controls like this Select widget. In the future we can also use them for auto-complete lists and tool tips.

https://user-images.githubusercontent.com/554369/236685193-0f251785-0a0a-4f53-a055-41e7f58dc64b.mov

### CODE REVIEW

File: src/textual/message.py [hunk 1]
```diff
@@ -38,6 +38,9 @@ class Message:
     namespace: ClassVar[str] = ""  # Namespace to disambiguate messages
 
     def __init__(self) -> None:
+        self.__post_init__()
```
Comment (willmcgugan) [inline src/textual/message.py:41]: This allows creating messages from dataclasses.

File: docs/widgets/select.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+A Select widget is a compact control to allow the user to select between a number of possible options.
+
+
+- [X] Focusable
+- [ ] Container
+
+
+The options in a select control may be passed in to the constructor or set later with [set_options][textual.widgets.Select.set_options].
+Options should be given as sequence of tuples consisting of two values: the first is the string (or [Rich Renderable](https://rich.readthedocs.io/en/latest/protocol.html)) to display in the control and list of options, the second is the value of option.
```
Comment (davep) [inline docs/widgets/select.md:13]:
```suggestion
Options should be given as a sequence of tuples consisting of two values: the first is the string (or [Rich Renderable](https://rich.readthedocs.io/en/latest/protocol.html)) to display in the control and list of options, the second is the value of option.
```

File: src/textual/binding.py [hunk 3]
```diff
@@ -74,7 +74,7 @@ def make_bindings(bindings: Iterable[BindingType]) -> Iterable[Binding]:
             for binding in bindings:
                 # If it's a tuple of length 3, convert into a Binding first
                 if isinstance(binding, tuple):
-                    if len(binding) != 3:
+                    if len(binding) not in (2, 3):
                         raise BindingError(
                             f"BINDINGS must contain a tuple of three strings, not {binding!r}"
```
Comment (davep) [inline src/textual/binding.py:79]: This error message needs an update.

File: src/textual/css/_styles_builder.py [hunk 4]
```diff
@@ -1003,6 +1005,30 @@ def process_grid_size(self, name: str, tokens: list[Token]) -> None:
         else:
             self.error(name, tokens[0], "expected two integers here")
 
+    def process_overlay(self, name: str, tokens: list[Token]) -> None:
```
Comment (davep) [inline src/textual/css/_styles_builder.py:1008]: Missing a docstring.

File: src/textual/css/_styles_builder.py [hunk 5]
```diff
@@ -1003,6 +1005,30 @@ def process_grid_size(self, name: str, tokens: list[Token]) -> None:
...
+        except StyleValueError:
+            self.error(
+                name,
+                tokens[0],
+                string_enum_help_text(name, VALID_OVERLAY, context="css"),
+            )
+        else:
+            self.styles._rules[name] = value  # type: ignore
+
+    def process_constrain(self, name: str, tokens: list[Token]) -> None:
```
Comment (davep) [inline src/textual/css/_styles_builder.py:1020]: Missing a docstring.

File: src/textual/widgets/_select.py [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+        box-sizing: content-box;
+        width: 1;
+        height: 1;
+        padding: 0 0 0 1;
+        color: $text-muted;
+        background: transparent;
+    }
+    """
+
+    has_value = var(False)
```
Comment (davep) [inline src/textual/widgets/_select.py:115]: I believe this should be typed as it helps the docs?

File: src/textual/widgets/_select.py [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+    A Select displays the current selection.
+    When activated with ++enter++ the widget displays an overlay with a list of all possible options.
+
+    """
+
+    BINDINGS = [("enter", "show_overlay")]
+    """
+    | Key(s) | Description |
+    | :- | :- |
+    | enter | Activate the overlay |
```
Comment (davep) [inline src/textual/widgets/_select.py:176]: <kbd>Space</kbd> is another common keystroke to active such a control (see macOS for example). I think I'd be tempted to support that too, perhaps even down arrow as well.

File: src/textual/widgets/_select.py [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+    Select.-expanded > SelectOverlay {
+        display: block;
+    }
+
+    Select.-expanded > SelectCurrent {
+        border: tall $accent;
+    }
+    """
+
+    expanded = var(False, init=False)
```
Comment (davep) [inline src/textual/widgets/_select.py:222]: I suspect this needs a type to help the docs?

File: src/textual/widgets/_select.py [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+    """The value of the select."""
+
+    class Changed(Message, bubble=True):
+        """Posted when the select value was changed.
+
+        This message can be handled using a `on_select_changed` method.
+
+        """
+
+        def __init__(
```
Comment (davep) [inline src/textual/widgets/_select.py:236]: Missing docstring.

File: src/textual/widgets/_select.py [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+        event.stop()
+        self.expanded = not self.expanded
+
+    @on(SelectOverlay.Dismiss)
+    def _select_overlay_dismiss(self, event: SelectOverlay.Dismiss) -> None:
+        """Dismiss the overlay."""
+        event.stop()
+        self.expanded = False
+        if not event.lost_focus:
+            """If the overlay didn't lose focus, we want to re-focus the select."""
```
Comment (davep) [inline src/textual/widgets/_select.py:354]: String used as a comment? (some folk do, I know, but not usual for us).

File: src/textual/widgets/_select.py
Comment (davep) [inline src/textual/widgets/_select.py:1]: There's one slightly surprising thing about how this works, when it comes to changing focus. I can see why it works this way but I wonder if we can work around it? If the list is dropped down and you're active in the list and you press <kbd>tab</kbd> it closes the dropdown list and moves focus to the next control.

On the other hand, in the same situation, if you press <kbd>shift</kbd>+<kbd>tab</kbd> it closes the dropdown list and "remains" on the same control. Of course, really, it *is* moving to the previous control in the focus chain but I feel it should ideally move to the control that's prior to the dropdown's `SelectCurrent`.

**Review by davep — COMMENTED** [review]
Second coffee of the morning read-through and code play, couple of questions.

File: src/textual/widgets/_select.py
Comment (willmcgugan) [reply src/textual/widgets/_select.py:1]: I'll think we'll live with that for now. Can't think of a simple workaround.

File: src/textual/css/_styles_builder.py (hunk 4 above)
Comment (willmcgugan) [reply src/textual/css/_styles_builder.py:1008]: None of the process methods do. Will leave it until we add all of them.

**Review by davep — APPROVED** [review]
Main questions resolved, looks good to go for me. Looking forward to making use of this.

I'm actually really taken by how good and smooth it all looks.

**Pull request merged**